src/lambdaconnect_sync/push.clj

```diff
@@ -418,41 +418,42 @@
-    :else []))
-
-(defn update-protected-fields [l-m r-m l r]
-  (cond
-    ; de Morgan's law:
-    (and (seq l) (seq r)) (vec (intersection (into #{} l) (into #{} r)))
-    (not= l-m r-m) (vec (set (concat l r)))
-    :else []))
-
-(defn- merge-fields [perms entity]
-  (let [{w-f :writable-fields p-f :protected-fields} perms
-        all-fields (into #{} (concat (keys (:attributes entity)) (keys (:relationships entity))))
-        all-fields-except-protected (difference all-fields (into #{} p-f))
-        writable-fields (vec (set (concat all-fields-except-protected w-f)))]
-    (if (and w-f p-f)
-      (-> perms
-          (dissoc :protected-fields)
-          (assoc :writable-fields writable-fields))
-      perms)))
-
-(defn- merge-permissions [permissions]
-  (reduce (fn [perms {c :create m :modify w :writable-fields p :protected-fields}]
-            (let [p-m (:modify perms)]
-              (-> perms
-                  (update :create (partial update-boolean-permissions c))
-                  (update :modify (partial update-boolean-permissions m))
-                  (update :writable-fields (partial update-writable-fields p-m m w))
-                  (update :protected-fields (partial update-protected-fields p-m m p))))) (first permissions) permissions))
-
+(spec/def ::writable-fields (spec/coll-of string?))
+(spec/def ::protected-fields (spec/coll-of string?))
+(spec/def ::modify boolean?)
+(spec/def ::create boolean?)
+
+(spec/def ::permission (spec/keys :req-un [::modify ::create] :opt-un [::writable-fields ::protected-fields]))
+
+(defn- merge-permissions [permissions entity]
+  {:pre [(spec/valid? (spec/coll-of ::permission) permissions) (spec/valid? :lambdaconnect-model.data-xml/entity entity)]
+   :post [(spec/valid? ::permission %)]}
+  (let [all-fields (into #{} (concat (keys (:attributes entity)) (keys (:relationships entity))))
+        ;; Simplified permissions always have writable-fields only.
+        simplified-permissions (map (fn [{:keys [writable-fields protected-fields modify] 
+                                          :as permission}]
+                                      (when writable-fields
+                                        (assert (subset? (set writable-fields) all-fields)
+                                                (str (set writable-fields) " is not a subset of " all-fields)))
+                                      (when protected-fields
+                                        (assert (subset? (set protected-fields) all-fields)
+                                                (str (set protected-fields) " is not a subset of " all-fields)))
+                                      (-> permission
+                                          (select-keys [:create :modify])
+                                          (assoc :writable-fields 
+                                                 (cond 
+                                                   (not modify) #{}
```

Comment:
Opcjonalnie: jeśli `modify` jest ustawione na false, a `protected-fields` lub `writable-fields` nie są puste, to dodałbym warning/assert, że `protected-fields` i `writable-fields` są ignorowane